src/analytics/base_token.rs

```diff
@@ -0,0 +1,54 @@
+// Copyright 2023 IOTA Stiftung
+// SPDX-License-Identifier: Apache-2.0
+
+use std::collections::HashMap;
+
+use super::TransactionAnalytics;
+use crate::types::{
+    ledger::{LedgerOutput, LedgerSpent},
+    stardust::block::{Address, Output},
+    tangle::MilestoneIndex,
+};
+
+/// Measures activity of the base token, such as Shimmer or IOTA.
+#[derive(Clone, Debug, Default)]
+pub struct BaseTokenActivity {
+    /// Represents the amount of tokens transfered. Tokens that are send back to an address are not counted.
+    pub booked_value: usize,
+    /// Represents the total amount of tokens transfered, independent of wether tokens were sent back to same address.
+    pub transferred_value: usize,
+}
+
+struct BaseTokenActivityAnalytics {
+    measurement: BaseTokenActivity,
+}
+
+impl TransactionAnalytics for BaseTokenActivityAnalytics {
+    type Measurement = BaseTokenActivity;
+
+    fn begin_milestone(&mut self, _: MilestoneIndex) {
+        self.measurement = BaseTokenActivity::default();
+    }
+
+    fn handle_transaction(&mut self, inputs: &[LedgerSpent], outputs: &[LedgerOutput]) {
+        let mut outflows: HashMap<&Address, usize> = HashMap::new();
+        for input in inputs {
+            if let Some(address) = input.output.output.owning_address() {
+                self.measurement.transferred_value += input.output.output.amount().0 as usize;
+                *outflows.entry(address).or_default() += input.output.output.amount().0 as usize;
```

Comment:
We should probably define helper functions to avoid `output.output.amount` weirdness.

Reply:
I agree. That or maybe we consider refactoring to `inner`. I think reading `input.output.inner.amount` would be fine. (reading `input.output.....` is also a bit weird I guess :smile:

Reply:
Obviously we should change `input` to `output` so that it's `output.output.output.amount`.